tango/integrations/torch/eval.py

```diff
@@ -0,0 +1,209 @@
+import random
+from collections import defaultdict
+from itertools import islice
+from typing import Dict, List, Optional, Sequence, Union
+
+import numpy as np
+import torch
+
+from tango.common.dataset_dict import DatasetDictBase
+from tango.common.exceptions import ConfigurationError
+from tango.common.lazy import Lazy
+from tango.common.tqdm import Tqdm
+from tango.format import Format, JsonFormat
+from tango.step import Step
+
+from .data import DataLoader
+from .eval_callback import EvalCallback
+from .model import Model
+from .util import check_dataset, move_to_device
+
+
+@Step.register("torch::eval")
+class TorchEvalStep(Step):
+    """
+    A basic PyTorch evaluation loop that pairs well with :class:`TorchTrainStep`.
+
+    .. tip::
+
+        Registered as a :class:`~tango.step.Step` under the name "torch::eval".
+
+    .. warning::
+
+        By default the metrics specified by the ``metric_names`` parameter
+        are aggregated by simply averaging across batches.
+        This behavior is usually correct for metrics like "loss" or "accuracy",
+        for example, but may not be correct for other metrics like "F1".
+
+        If this is not correct for your metric you will need to handle the aggregation
+        internally in your model or with an :class:`EvalCallback`
+        using the :meth:`EvalCallback.post_batch()` method.
+        Then set the parameter ``auto_aggregate_metrics`` to ``False``.
+
+    """
+
+    DETERMINISTIC = True
+    CACHEABLE = True
+    FORMAT: Format = JsonFormat()
```

Comment:
MyPy always complains about this one if you don't annotate `FORMAT`. Idk why.

Reply:
I have noticed this, but I have long ago given up trying to understand mypy...